src/qwircl/core.cljs

```diff
@@ -78,14 +83,78 @@
 
 (defn empty-location? [[x y] grid]
   (let [location (get-in grid [x y])]
-    (and 
+    (or
+     true
      (nil? (:color location))
      (nil? (:shape location)))))
 
-(defn valid-play? [state clicked]
-  (and 
-   (empty-location? clicked (:grid state))
-   true))
+(defn same-line? [positions [x y]]
+  (or (empty? positions)
+      (every? #(= (first (:coordinates %)) x)
+              positions)
+      (every? #(= (second (:coordinates %)) y) 
+              positions)))
+
+(def transformations {:horizontal {:x [dec inc] :y [identity identity]}
+                      :vertical {:x [identity identity] :y [dec inc]}})
+(defn reduce-neighbors 
```

Comment:
I think this can be broken up into two steps:
1. finding all the tiles along a line given  a direction and position: `(get-line grid position direction)`
2. plain old reduce for validating properties of the tiles: `(reduce (same-color-pred (first line)) (rest line))` etc